### PR TITLE
Adds a new "migrating from 1.x to 2.x" section to the documentation

### PR DESCRIPTION
**Pull request by ajsantander**

### CODE REVIEW

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+---
+id: migrate_from_1.x_to_2.x
+title: Migrating to 2.x
+---
+
+In this section, you'll find everything you need to know if you're coming from ZeppelinOS 1.x and want to start using 2.x.
+
+## Terminology and fundamentals
+
+2.x's most fundamental change is that it expands on zOS's linking features. In 1.x, you were able to link your projects to what used to be called an "on-chain standard library". This "stdlib" was basically a pre-deployed version of OpenZeppelin, which gave your projects instant access to hundreds of standard implementations without having to compile or deploy any code whatsoever. This proved to be a very powerful, innovative concept, but we realized that it could be taken much further... In 1.x, you could connect to different versions of this "standard library" but there was just this ONE "standard library" that your application could link to. 2.x takes this to a whole new level; you can now connect your project to multiple "standard libraries", but that's not all, this makes the entire concept of what used to be a "standard library" evolve to the concept of an "EVM package".
```
Comment (spalladino) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: > This "stdlib" was basically a pre-deployed version of OpenZeppelin

Not sure we want to frame it like this. Even though the only stdlib available _was_ OZ, the intention was to have different competing stdlibs. I'd rephrase it as "a pre-deployed version of a set of common reusable contracts, like OZ", or similar.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+---
+id: migrate_from_1.x_to_2.x
+title: Migrating to 2.x
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: Maybe we can call it: What's new in ZeppelinOS 2.0?

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+In 2.x by default, your zOS project will not use any of its App or Package on-chain architecture unless you explicitly state that your code is meant to be reusable as an EVM package. When you init a zOS project, there is no architecture other than your project's code. ZeppelinOS simply uses the CLI to manage proxies for you. Now, if you intend your project to be a reusable EVM package, you can `publish` your code, and the CLI will seamlessly deploy the necessary App and Package architecture which will allow your code to exist as a discoverable deployed EVM package.
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: you can `publish` your code -> you can run `zos publish` on your project

Comment (elopio) [reply packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: the necessary App and Package architecture -> the necessary contracts
a discoverable deployed EVM package -> an EVM package.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+## New commands
+
+2.x introduces a few new commands to its CLI which are described below:
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: The commands reference will be in https://docs.zeppelinos.org/docs/cli_main.html
so, I would remove the details, and just link to that page.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+publishes your project to the selected network
+
+Options:
+  -n, --network <network>  network to be used
+  -f, --from <from>        specify transaction sender address
+  --timeout <timeout>      timeout in seconds for each blockchain transaction (defaults to 600s)
+  -h, --help               output usage information
+```
+
+As explained in the ["Terminology and fundamentals" section](https://docs.zeppelinos.org/docs/migrate_from_1.x_to_2.x.html#terminology-and-fundamentals), this command is what transforms your project to a reusable EVM package. It deploys the App and Package architecture and ensures that your code is discoverable as an EVM package.
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: This paragraph and the next two should be on the guide that explains the push command.
It is WIP on my restructure task, so I can now just copy :)

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+Usage: check [contract] [options]
+
+checks your contracts for potential issues
+
+Options:
+  --skip-compile  skips contract compilation
+  -h, --help      output usage information
+```
+
+As discussed in the ["Safety checks" section of the documentation](https://docs.zeppelinos.org/docs/advanced.html#safety-checks), the zOS CLI performs a series of safety checks on your contracts as a sub-routine of some of it's commands. The new `check` command allows you to perform these checks in a standalone manner.
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: Maybe, the command help should print a link to that page :thinking:

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+Opposite to the `link` command.
+
+## Modified commands
+
+In 2.x, there are minor changes to the commands from 1.x, like some of them now having options like `--skip-compile`, but there are also more signiffcant changes.
+
+*`bump`* is now only relevant for published EVM packages, and is no longer applicable to projects that are not intended for reusability.
+
+*`link`* can now be called for multiple EVM packages.
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: maybe: with multiple EVM packages as arguments.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+## Compatibility
+
+Unfortunately, 2.x introduces breaking changes, so a project that was created using 1.x cannot be upgraded to 2.x. This means that if you upgrade your global `zos` NPM package to 2.x, it won't be compatible with projects that were created using the old version of `zos`. In such cases, the CLI will detect the incompatibility and warn you. This is one of the reasons why `npx zos` is recommended over `zos` usage, so that each NPM project can target its own `zos` version.
+
+#### Changes to your contracts
+
+When it comes to Solidity code, there is nothing special you need to consider when using zOS 2.x other than using initializers instead of constructors. For more info on this, see the ["Initializers vs constructors"](https://docs.zeppelinos.org/docs/advanced.html#initializers-vs-constructors) section of the documentation.
+
+As with 1.x, zOS 2.x will manage all upgradeability and package linking features without you having to any use of special Solidity syntax. In this sense, zOS is completely language agnostic.
```
Comment (elopio) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: having to any use of -> having to use

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md (hunk 2 above)
Comment (spalladino) [reply packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: +1! Besides, we'll eventually write an actual migration guide, as we script the process

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+What is an EVM package? It is basically a piece of reusable code that has been deployed to the blockchain in the form of EVM bytecode. Anyone can create an EVM package, and your zOS projects will be able to connect to a vast ecosystem of pre-compiled, pre-deployed reusable code in the form of EVM packages.
+
+Similarly, as we realized what an EVM package was, we realized what it's not. In 1.x, your zOS project's contracts were always managed by two contracts: App and Package. This made your contracts upgradeable and gave them the ability to link to EVM packages (see [Contract architecture](https://docs.zeppelinos.org/docs/architecture.html) for more info). We realized that these two contracts made sense when it came to EVM packages, but didn't when it came to parts of your project that weren't meant to be reusable code, i.e. code that just needed to be upgradeable but not reusable. And so, in 2.x, upgradeability, and the ability to link to EVM packages were completely separated.
```
Comment (spalladino) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: >  We realized that these two contracts made sense when it came to EVM packages, but didn't when it came to parts of your project that weren't meant to be reusable code

This is true for Package and ImplementationProvider, but not for App. An App is not needed for sharing or reusing code; it only concentrates ownership of all proxies in a single address, different from the user's account. This has two benefits: 1) no issues when interacting with transparent proxies from the user's account, and 2) you can transfer ownership of the app to a multisig or another governance structure, to decentralize upgradeability control of all your contracts.

Comment (spalladino) [reply packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: After writing this, I realize that it could be somewhat confusing that `publish` has this twofold behaviour. We'll need to be particularly careful on how we explain it. And eventually, we could [remove one of those behaviours](https://github.com/zeppelinos/zos/issues/291) and just keep the _publishing reusable code_ part.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+## Modified commands
+
+In 2.x, there are minor changes to the commands from 1.x, like some of them now having options like `--skip-compile`, but there are also more signiffcant changes.
+
+*`bump`* is now only relevant for published EVM packages, and is no longer applicable to projects that are not intended for reusability.
+
+*`link`* can now be called for multiple EVM packages.
+
+*`init`* no longer has the `--lib` option.
```
Comment (spalladino) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: We should explain the rationale for this in the fundamentals section. An option would be, in the section where we explain how to publish, add a comparison to 1.x in which you had to state from the very beginning (ie when running `init`) where the project would be a reusable package or not.

> Now, if you intend your project to be a reusable EVM package, you can `publish` your code, and the CLI will seamlessly deploy the necessary App and Package architecture which will allow your code to exist as a discoverable deployed EVM package.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+## Compatibility
+
+Unfortunately, 2.x introduces breaking changes, so a project that was created using 1.x cannot be upgraded to 2.x. This means that if you upgrade your global `zos` NPM package to 2.x, it won't be compatible with projects that were created using the old version of `zos`. In such cases, the CLI will detect the incompatibility and warn you. This is one of the reasons why `npx zos` is recommended over `zos` usage, so that each NPM project can target its own `zos` version.
```
Comment (spalladino) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: > that was created using 1.x cannot be upgraded to 2.x

Cannot be _automatically_ upgraded to 2.x. There is a good chance that we'll set up a manual migration guide/script in the near future.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+#### Changes to your contracts
+
+When it comes to Solidity code, there is nothing special you need to consider when using zOS 2.x other than using initializers instead of constructors. For more info on this, see the ["Initializers vs constructors"](https://docs.zeppelinos.org/docs/advanced.html#initializers-vs-constructors) section of the documentation.
```
Comment (spalladino) [inline packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: I'd add here that the `Migratable` contract has been deprecated, and we are shifting to a much simpler flavour of `Initializable`.

File: packages/docs/docs/docs/migrate_from_1.x_to_2.x.md (hunk 8 above)
Comment (spalladino) [reply packages/docs/docs/docs/migrate_from_1.x_to_2.x.md]: > zOS is completely language agnostic

I'd avoid making these claims. The CLI is very closely tied to Solidity, and lib is not that decoupled either, even if the upgradeability and linking patterns themselves *are* language agnostic.

**Comment by spalladino** [comment]
Just left a few individual comments (don't know why I didn't go with a review instead) on a few details that were missing, but the guide is great work @ajsantander!! Kudos :+1:
